### ISSUE TITLE
False positive warning: redundant specification of type arguments

### ISSUE THREAD

**Issue by alexcormier** [eclipse-jdt/eclipse.jdt.core#1326]
Given the following classes:

```java
class Inner<T> {}
class Outer<T> {
   Outer(T t) {}
   Outer<T> self() { return this; }
}
```

This code generates a warning:

```java
Outer<Inner<String>> x = new Outer<>(new Inner<String>()).self();
//                                       ^^^^^
// Redundant specification of type arguments <String>
```

But omitting the type argument generates an error as type inference fails and falls-back to `Object`:

```java
Outer<Inner<String>> x = new Outer<>(new Inner<>()).self();
//                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// Type mismatch: cannot convert from Outer<Inner<Object>> to Outer<Inner<String>>
```

This is a relatively recent change that I assume was introduced with #1036.

**Comment by stephan-herrmann** [comment]
This could indeed be caused by this change of strategy in #1036:

@jarthana:
> For method invocations, the allocation expression does not get its expected type soon enough. So, we should wait till it's available.

It might look as if the expected type for `new Inner<_>()` would come from the inferred outer `Outer.<init>(String)`, but that's circular reasoning. Without the explicit type argument in `new Inner<String>()` the outer diamond would indeed infer `<Object>` because this allocation expression is a _standalone_ expression that cannot see passed the dot of `.self()`.

**Comment by jarthana** [comment]
> This could indeed be caused by this change of strategy in #1036:
> 
> @jarthana:
> 
> > For method invocations, the allocation expression does not get its expected type soon enough. So, we should wait till it's available.
> 
> It might look as if the expected type for `new Inner<_>()` would come from the inferred outer `Outer.<init>(String)`, but that's circular reasoning. Without the explicit type argument in `new Inner<String>()` the outer diamond would indeed infer `<Object>` because this allocation expression is a _standalone_ expression that cannot see passed the dot of `.self()`.

OK, I see that the original fix is not helping here and perhaps can be reverted. But I don't really know how else to proceed with that issue. Do you have some tips/hints for me?

**Comment by stephan-herrmann** [comment]
> OK, I see that the original fix is not helping here and perhaps can be reverted.

I actually found it convenient to keep it and just slightly modify (so I didn't have to fully understand #1036 :wink:), although the resulting machinery just for this one warning is now spread over more locations than one might like.

> But I don't really know how else to proceed with that issue. Do you have some tips/hints for me?

I didn't have good hints before doing some experiments in the debugger, but then it all seemed to fall in place.

WDYT?

**Comment by stephan-herrmann** [comment]
[Test failures](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1378#issuecomment-1722434904) remind me that the fix is effective only at compliance 1.8+. As the issue's topic is diamond expressions, I placed the test case into class GenericsRegressionTest_1_7 which promises to be applicable to 1.7+. 

I'll take a quick look, if a corresponding fix can also be made for 1.7, otherwise, and since 1.7 is not a top priority any longer, I will simply disable the new tests below 1.8.
